macOS/unwindmacos.cpp

```diff
@@ -20,49 +20,116 @@ limitations under the License.
 #include <mach-o/dyld_images.h>
 #include <mach-o/nlist.h>
 
-UnwindDataMacOS::UnwindDataMacOS() {
-  unwind_section_address = NULL;
-  unwind_section_size = 0;
-  unwind_section_buffer = NULL;
-  unwind_section_header = NULL;
-  last_encoding_lookup = LastEncodingLookup();
-  have_data_to_register = false;
-}
+#include <third_party/llvm/libunwind/dwarf2.h>
+#include <third_party/llvm/libunwind/CompactUnwinder.hpp>
 
-UnwindDataMacOS::~UnwindDataMacOS() {
-  free(unwind_section_buffer);
-  unwind_section_buffer = NULL;
-}
+constexpr const unsigned char UnwindGeneratorMacOS::register_assembly_x86[];
 
-void UnwindDataMacOS::AddEncoding(size_t original_address, size_t translated_address) {
+void UnwindDataMacOS::LookupEncoding(size_t original_address) {
   if (encoding_map.empty()) {
+    last_lookup = LastLookup();
     return;
   }
 
-  if (original_address < last_encoding_lookup.original_min_address
-      || last_encoding_lookup.original_max_address <= original_address) {
+  if (last_lookup.IsEncodingMiss(original_address)) {
+    last_lookup = LastLookup();
     auto it = encoding_map.upper_bound(original_address);
     if (it == encoding_map.begin()) {
-      last_encoding_lookup = LastEncodingLookup(0, 0, it->first - 1);
+      last_lookup.SetEncoding(0, 0, it->first - 1);
     } else if (it == encoding_map.end()) { // Sentinel entry
-      last_encoding_lookup = LastEncodingLookup(0, prev(it)->first, -1);
+      last_lookup.SetEncoding(0, prev(it)->first, -1);
+    } else {
+      last_lookup.SetEncoding(prev(it)->second, prev(it)->first, it->first - 1);
+    }
+  }
+}
+
+void UnwindDataMacOS::LookupLSDA(size_t original_address) {
+  if (last_lookup.IsLsdaMiss(original_address)) {
+    if (!lsda_map.empty() && last_lookup.encoding & UNWIND_HAS_LSDA) {
```

Comment:
A condition of `x && y & z` is a bit hard to read even if you know the evaluation order of the operators. I'd put `last_lookup.encoding & UNWIND_HAS_LSDA` in parenthesis here.